python/ray/ml/trainer.py

```diff
@@ -210,8 +241,74 @@ def fit(self) -> Result:
             TrainingFailedError: If any failures during the execution of
             ``self.as_trainable()``.
         """
-        raise NotImplementedError
+        trainable = self.as_trainable()
+
+        from ray import tune
+        from ray.tune import TuneError
+
+        # TODO(amog/xwjiang): Replace with Tuner.
+        try:
+            analysis = tune.run(run_or_experiment=trainable, **self.run_config)
+        except TuneError:
+            raise TrainingFailedError
+        else:
+            assert len(analysis.trials) == 1
+
+            trial = analysis.trials[0]
+
+            result = Result(
+                metrics=trial.last_result,
+                checkpoint=Checkpoint.from_directory(trial.checkpoint.value)
+                if trial.checkpoint.value
+                else None,
+            )
+
+            return result
 
     def as_trainable(self) -> Type[Trainable]:
         """Convert self to a ``tune.Trainable`` class."""
-        raise NotImplementedError
+
+        base_config = self._param_dict
+        trainer_cls = self.__class__
+
+        def train_func(config, checkpoint_dir=None):
+            # config already contains merged values.
+            # Instantiate new Trainer in Trainable.
+            trainer = trainer_cls(**config)
+
+            if checkpoint_dir:
+                trainer.resume_from_checkpoint = Checkpoint.from_directory(
+                    checkpoint_dir
+                )
+
+            trainer.setup()
+            trainer.preprocess_datasets()
+            trainer.training_loop()
+
+        trainable_cls = wrap_function(train_func)
+
+        scaling_config = self.scaling_config
+
+        class TrainTrainable(trainable_cls):
+            """Add default resources to the Trainable."""
+
+            def __init__(self, *args, **kwargs):
+                super().__init__(*args, **kwargs)
+
+                # Create a new config by merging the dicts.
+                self._merged_config = merge_dicts(base_config, self.config)
+
+            def _trainable_func(self, config, reporter, checkpoint_dir):
+                # We ignore the config passed by Tune and instead use the merged
+                # config which includes the initial Trainer args.
+                super()._trainable_func(self._merged_config, reporter, checkpoint_dir)
+
+            @classmethod
+            def default_resource_request(cls, config):
+                updated_scaling_config = config.get("scaling_config", scaling_config)
```

Comment:
For my understanding, when do we ever need to override this with an input `config`? Why not just redefine the `Trainer`'s `scaling_config` instead?

Reply:
This is the config that's coming from tune.run/Tuner.fit, so we want to override scaling config if it's included as a hyperparameter.